token-lending/program/src/processor.rs

```diff
@@ -1528,6 +1546,195 @@ fn process_set_lending_market_owner(
+    }
+
+    if flash_loan_end_instruction.accounts[1].pubkey != *reserve_liquidity_account_info.key {
+        msg!("Different reserve liquidity supply between flash loan start and flash loan end");
+        return Err(LendingError::InvalidFlashLoanEnd.into());
+    }
+
+    if flash_loan_end_instruction.accounts[2].pubkey != *lending_market_account_info.key {
+        msg!("Different lending market between flash loan start and flash loan end");
+        return Err(LendingError::InvalidFlashLoanEnd.into());
+    }
+
+    let authority_signer_seeds = &[
+        lending_market_account_info.key.as_ref(),
+        &[lending_market.bump_seed],
+    ];
+    let lending_market_authority_pubkey =
+        Pubkey::create_program_address(authority_signer_seeds, program_id)?;
+
+    if derived_lending_market_account_info.key != &lending_market_authority_pubkey {
+        return Err(LendingError::InvalidMarketAuthority.into());
+    }
+
+    reserve.liquidity.flash_loan_start(liquidity_amount)?;
+    Reserve::pack(reserve, &mut reserve_account_info.data.borrow_mut())?;
+
+    spl_token_transfer(TokenTransferParams {
+        source: reserve_liquidity_account_info.clone(),
+        destination: destination_token_account_info.clone(),
+        amount: liquidity_amount,
+        authority: derived_lending_market_account_info.clone(),
+        authority_signer_seeds,
+        token_program: token_account.clone(),
+    })?;
+
+    Ok(())
+}
+
+fn process_flash_loan_end(program_id: &Pubkey, accounts: &[AccountInfo]) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+    let reserve_account_info = next_account_info(account_info_iter)?;
+    let reserve_liquidity_supply = next_account_info(account_info_iter)?;
+    let lending_market_account_info = next_account_info(account_info_iter)?;
+    let derived_lending_market_account_info = next_account_info(account_info_iter)?;
+    let flash_loan_fees_account_info = next_account_info(account_info_iter)?;
+    let token_program_id = next_account_info(account_info_iter)?;
+
+    let lending_market = LendingMarket::unpack(&lending_market_account_info.data.borrow())?;
+
+    if reserve_account_info.owner != program_id {
+        return Err(LendingError::InvalidAccountOwner.into());
+    }
+
+    let mut reserve = Reserve::unpack(&reserve_account_info.data.borrow())?;
+    if &reserve.lending_market != lending_market_account_info.key {
+        msg!("Invalid reserve lending market account");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+
+    let liquidity_supply_token_account =
+        Account::unpack_from_slice(&reserve_liquidity_supply.try_borrow_data()?)?;
+    let (origination_fee, host_fee) = reserve
+        .config
+        .fees
+        .calculate_flash_loan_fees(reserve.liquidity.flash_loaned_amount)?;
+
+    if reserve.liquidity.available_amount + reserve.liquidity.flash_loaned_amount + origination_fee
+        > liquidity_supply_token_account.amount
```

Comment:
Note that if anyone ever mistakenly sends tokens to the liquidity supply token account, flash loans will be able to take the extra funds because the amount they need to pay back is a bit smaller. I *think* this is ok, but I think I would feel better if `FlashLoanEnd` did the token transfer for the user.

Reply:
I have changed to `FlashLoanEnd` do the transfer for the user.